### PR TITLE
Account System overhaul. Addresses #76.

### PR DESCRIPTION
**Pull request by creatorfromhell**
This PR addresses a couple of issues.

First, it addresses the issue that BankAccount is misidentified as accounts that are only suitable for bank-related functions when they should be utilized for any account that is not a PlayerAccount. That's why they get renamed to GenericAccounts in order to create a clear inference that they are anything that is not a player account.

Secondly, this addresses the fact that not all accounts should be associated with a UUID. While it is true that the odds of two UUIDs being generated are the same, the fact remains that most economy solutions treat non-player accounts without a UUID. For instance, towny, kingdoms, and lands accounts are prefixed-Name. For instance, a town account is town-TownName in towny. This is something that should be kept in mind, as it has also been utilized as the standard for other platforms such as Sponge. If this truly is supposed to be a platform-independent solution then thought needs to be put into making it as flexible as possible, which in Minecraft terms means players have a UUID and non-players do not.

Lastly, and most insignificant, I changed CONSULT permissions to BALANCE, because why overcomplicate it? It should be straightforward.

### CODE REVIEW

**Review by lokka30 — COMMENTED** [review]
I am very unsure about this change.

I am tilting on agreeing with:
* Making `GenericAccount`s based on String identifiers instead of UUIDs. Previously all the economy providers needed to do was have a database mapping the two but I guess it's okay to switch to Strings.

I disagree with;
* The name `GenericAccount`: it could wrongly suggest that the class is a generic class. Previously, Treasury named `NonPlayerAccount` - perhaps this is better?
* The removal of bank-specific methods in the EconomyProvider file. Player and bank/non-player accounts are created differently
* 'Bank Account permissions' now being 'Account permissions' - they're meant to be for just banks


Spotted issues:
* `BankAccountTransactionEvent` should be `GenericAccountTransactionEvent` in this case

I agree with the renaming of `CONSULT` though :+1:

Thanks for the PR :)

**Comment by creatorfromhell** [comment]
> I am very unsure about this change.
> 
> I am tilting on agreeing with:
> 
> * Making `GenericAccount`s based on String identifiers instead of UUIDs. Previously all the economy providers needed to do was have a database mapping the two but I guess it's okay to switch to Strings.
> 
> I disagree with;
> 
> * The name `GenericAccount`: it could wrongly suggest that the class is a generic class. Previously, Treasury named `NonPlayerAccount` - perhaps this is better?
> * The removal of bank-specific methods in the EconomyProvider file. Player and bank/non-player accounts are created differently
> * 'Bank Account permissions' now being 'Account permissions' - they're meant to be for just banks
> 
> Spotted issues:
> 
> * `BankAccountTransactionEvent` should be `GenericAccountTransactionEvent` in this case
> 
> I agree with the renaming of `CONSULT` though 👍
> 
> Thanks for the PR :)

Here's the reasoning behind this. Having dedicated methods for bank accounts doesn't make sense, especially when all non player accounts are possible to have methods. I.e. town accounts have multiple methods, faction accounts, and this is why it shouldn't be restricted to a specific function.

**Comment by lokka30** [comment]
> > I am very unsure about this change.
> > I am tilting on agreeing with:
> > 
> > * Making `GenericAccount`s based on String identifiers instead of UUIDs. Previously all the economy providers needed to do was have a database mapping the two but I guess it's okay to switch to Strings.
> > 
> > I disagree with;
> > 
> > * The name `GenericAccount`: it could wrongly suggest that the class is a generic class. Previously, Treasury named `NonPlayerAccount` - perhaps this is better?
> > * The removal of bank-specific methods in the EconomyProvider file. Player and bank/non-player accounts are created differently
> > * 'Bank Account permissions' now being 'Account permissions' - they're meant to be for just banks
> > 
> > Spotted issues:
> > 
> > * `BankAccountTransactionEvent` should be `GenericAccountTransactionEvent` in this case
> > 
> > I agree with the renaming of `CONSULT` though 👍
> > Thanks for the PR :)
> 
> Here's the reasoning behind this. Having dedicated methods for bank accounts doesn't make sense, especially when all non player accounts are possible to have methods. I.e. town accounts have multiple methods, faction accounts, and this is why it shouldn't be restricted to a specific function.
> 
> Also when it comes to the removal of generic specific methods it only makes sense if an account is no different than a non player account it terms of identification there shouldn't be similar methods for both. This provides the ability for the economy provider to have new account types without worrying about how a consumer will work with them.

I'm still very uncertain about this. And, what did you think about my other comments? :)

I'll give it some time in case anyone wants to comment before I try make a deeper analysis on this.

**Comment by creatorfromhell** [comment]
> > > I am very unsure about this change.
> > > I am tilting on agreeing with:
> > > 
> > > * Making `GenericAccount`s based on String identifiers instead of UUIDs. Previously all the economy providers needed to do was have a database mapping the two but I guess it's okay to switch to Strings.
> > > 
> > > I disagree with;
> > > 
> > > * The name `GenericAccount`: it could wrongly suggest that the class is a generic class. Previously, Treasury named `NonPlayerAccount` - perhaps this is better?
> > > * The removal of bank-specific methods in the EconomyProvider file. Player and bank/non-player accounts are created differently
> > > * 'Bank Account permissions' now being 'Account permissions' - they're meant to be for just banks
> > > 
> > > Spotted issues:
> > > 
> > > * `BankAccountTransactionEvent` should be `GenericAccountTransactionEvent` in this case
> > > 
> > > I agree with the renaming of `CONSULT` though 👍
> > > Thanks for the PR :)
> > 
> > 
> > Here's the reasoning behind this. Having dedicated methods for bank accounts doesn't make sense, especially when all non player accounts are possible to have methods. I.e. town accounts have multiple methods, faction accounts, and this is why it shouldn't be restricted to a specific function.
> > Also when it comes to the removal of generic specific methods it only makes sense if an account is no different than a non player account it terms of identification there shouldn't be similar methods for both. This provides the ability for the economy provider to have new account types without worrying about how a consumer will work with them.
> 
> I'm still very uncertain about this. And, what did you think about my other comments? :)
> 
> I'll give it some time in case anyone wants to comment before I try make a deeper analysis on this.

The other comments are fair and I'll change accordingly. I still stand by the fact that account permissions should be allowed on all non player accounts, because this is the most flexible option that would allow for the most future proof system, in my opinion. Also, to further explain why it's better to return Account vs GenericAccount. If an economy provider wishes they could make new account classes such as TownAccount and return it without needing new methods. This allows for lots of flexibility when it comes to API design and limits restrictions.

**Review by MrIvanPlays — APPROVED** [review]
what lokka mentioned I agree with, otherwise LGTM

**Comment by MrNemo64** [comment]
> The other comments are fair and I'll change accordingly. I still stand by the fact that account permissions should be allowed on all non player accounts, because this is the most flexible option that would allow for the most future proof system, in my opinion. Also, to further explain why it's better to return Account vs GenericAccount. If an economy provider wishes they could make new account classes such as TownAccount and return it without needing new methods. This allows for lots of flexibility when it comes to API design and limits restrictions.

If we move the permissions methods to Account then GenericAccount is the same as Account meaning that we can remove it and we don't have the problem of the name

**Comment by lokka30** [comment]
> > The other comments are fair and I'll change accordingly. I still stand by the fact that account permissions should be allowed on all non player accounts, because this is the most flexible option that would allow for the most future proof system, in my opinion. Also, to further explain why it's better to return Account vs GenericAccount. If an economy provider wishes they could make new account classes such as TownAccount and return it without needing new methods. This allows for lots of flexibility when it comes to API design and limits restrictions.
> 
> If we move the permissions methods to Account then GenericAccount is the same as Account meaning that we can remove it and we don't have the problem of the name

I believe we should keep NonPlayerAccount even if it implements Account with no changes. It's still a specific type of account, but it doesn't add any features upon it.

**Review by lokka30 — APPROVED** [review]

**Comment by lokka30** [comment]
@Jikoo would you be able to check this PR? If not, no worries.

@creatorfromhell are we ready to merge this once Jikoo replies? Thank you very much for the PR!

**Comment by creatorfromhell** [comment]
> @Jikoo would you be able to check this PR? If not, no worries.
> 
> @creatorfromhell are we ready to merge this once Jikoo replies? Thank you very much for the PR!

Sure the only question left is Are the permissions methods staying in NonPlayerAccount or moving to Account to be able to be utilized in player accounts too?

**Comment by lokka30** [comment]
> > @Jikoo would you be able to check this PR? If not, no worries.
> > @creatorfromhell are we ready to merge this once Jikoo replies? Thank you very much for the PR!
> 
> Sure the only question left is Are the permissions methods staying in NonPlayerAccount or moving to Account to be able to be utilized in player accounts too?

I believe we can move the permissions to use on all accounts. Although, one thing that should be firm is that Player accounts have all permissions and one member only (being the player itself).

**Review by lokka30 — APPROVED** [review]

**Review by Jikoo — APPROVED** [review]
Beyond naming nitpick, looks good to me.

I don't particularly like the shift away from UUIDs. I understand that most systems tend to use UUIDs exclusively for players, but that doesn't mean they're correct to do so. Who cares if a player also has a unique identifier? All accounts should have a unique identifier, it just so happens that player accounts' identifiers should also match their owners' identifiers. Why make the identifier for other accounts a String when UUIDs exist? I guess the counterpoint is that the API user is still free to use UUIDs in string form like how Towny surely will, but it seems weird to use a separate construct for a unique ID when that is what a UUID is for. That was why we added the optional name, user friendliness.

**Comment by lokka30** [comment]
> Beyond naming nitpick, looks good to me.
> 
> I don't particularly like the shift away from UUIDs. I understand that most systems tend to use UUIDs exclusively for players, but that doesn't mean they're correct to do so. Who cares if a player also has a unique identifier? All accounts should have a unique identifier, it just so happens that player accounts' identifiers should also match their owners' identifiers. Why make the identifier for other accounts a String when UUIDs exist? I guess the counterpoint is that the API user is still free to use UUIDs in string form like how Towny surely will, but it seems weird to use a separate construct for a unique ID when that is what a UUID is for. That was why we added the optional name, user friendliness.

I have the exact same opinion. Although, I don't mind.

**Comment by MrNemo64** [comment]
> I believe we can move the permissions to use on all accounts. Although, one thing that should be firm is that Player accounts have all permissions and one member only (being the player itself).

At this point this is left for the provider to define. The player associated with the PlayerAccount will always have all permissions but depending on what the provider wants, other players may have permissions too and I don't think there is anything wrong with that.

File: api/src/main/java/me/lokka30/treasury/api/economy/account/NonPlayerAccount.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/account/NonPlayerAccount.java:15]: No longer called like that.
Since the NonPlayerAccount exists only to diferenciate player and non player accounts I think that should be mentioned on the description of the class

File: api/src/main/java/me/lokka30/treasury/api/economy/account/SharedAccountPermission.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/account/SharedAccountPermission.java:20]: By calling this `SharedAccountPermission` and being used by all bank accounts we are saying that PlayerAccounts are also shared accounts, whitch may or may not be true, it is left for the provider to specify.

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:87]: Owners no longer exist in the context of bank accounts. We miss the removal of this failure entry

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:79]: No longer called `BankAccount`

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:75]: No longer called `BankAccount`

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:83]: No longer called `BankAccount`

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:91]: No longer called `BankAccount`

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:95]: No longer called `BankAccount`

File: api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java
Comment (MrNemo64) [inline api/src/main/java/me/lokka30/treasury/api/economy/response/FailureReason.java:99]: No longer called `BankAccount`

File: core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java:97]: No longer called `GenericAccount`

File: core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java:33]: No longer called `GenericAccount`

File: core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java:38]: No longer called `GenericAccount`

File: core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/GenericAccountMigrator.java:43]: No longer called `GenericAccount`

**Review by MrNemo64 — CHANGES_REQUESTED** [review]
There are some places that Bank or Generic accounts are referenced. I tryed to mark them all but I may have mist some and there are parts of the code that I couldn't comment (the ones that didn't change)

EDIT: Also you didn't add yourself to the `@author` tag at the methods/classes

File: api/src/main/java/me/lokka30/treasury/api/economy/account/SharedAccountPermission.java
Comment (creatorfromhell) [reply api/src/main/java/me/lokka30/treasury/api/economy/account/SharedAccountPermission.java:20]: I think this naming scheme better suits it since the methods are in Account now.

Comment (MrIvanPlays) [inline api/src/main/java/me/lokka30/treasury/api/economy/account/SharedAccountPermission.java:20]: Maybe we can just name it AccountPermission? I'm fine with that name too but what about AccountPermission?

**Comment by MrIvanPlays** [comment]
also you may want to merge `master` into this branch in order to resolve conflicts as we currently cannot merge this. If you can't right now I will.

**Comment by MrIvanPlays** [comment]
Thanks!

**Pull request merged**
